### PR TITLE
Filter available runtimes based on configuration setting

### PR DESCRIPTION
**Pull request by kevin-bates**
This pull request allows operators to NOT expose runtime processors that might otherwise be available.  By default, all three out-of-the-box runtime processors ('local', 'airflow', 'kfp') are exposed because their entry points are provided as part of Elyra's packaging.  With this change, operators can select which runtime processors should be exposed by setting the command-line option `--PipelineProcessorRegistry.runtimes=<runtime>`.  If they wish to expose multiple runtimes, then additional `--PipelineProcessorRegistry.runtimes=<runtime>` options are required.

## CLI Configuration Option
For example, a command line of `jupyter elyra --PipelineProcessorRegistry.runtimes=airflow` will produce the following set of Elyra tiles:
Screenshot 2023-02-21 at 1 48 25 PM
and the following log entries:
```
[I 2023-02-21 13:47:58.149 ElyraApp] Registering airflow processor 'elyra.pipeline.airflow.processor_airflow.AirflowPipelineProcessor'...
[I 2023-02-21 13:47:58.834 ElyraApp] Although runtime 'kfp' is installed, it is not in the set of runtimes configured via '--PipelineProcessorRegistry.runtimes' and will not be available.
[I 2023-02-21 13:47:58.918 ElyraApp] Although runtime 'local' is installed, it is not in the set of runtimes configured via '--PipelineProcessorRegistry.runtimes' and will not be available.
```

Similarly, a command line of `jupyter elyra --PipelineProcessorRegistry.runtimes=kfp --PipelineProcessorRegistry.runtimes=local` will produce tiles:
Screenshot 2023-02-21 at 1 50 13 PM

---
## File-based Configuration Option
Note that these options can also be set in the `jupyter_elyra_config.py` configuration file:
```
#------------------------------------------------------------------------------
# PipelineProcessorRegistry(SingletonConfigurable) configuration
#------------------------------------------------------------------------------
## The runtimes to use during this Elyra instance.  (env
#  ELYRA_PROCESSOR_RUNTIMES) 
#  Default: []
c.PipelineProcessorRegistry.runtimes = ['local', 'airflow']
```
such that a command of `jupyter elyra` yields:
image

### What changes were proposed in this pull request?
- See above, plus
- Add new _Configuring the pipeline editor_ topic to the _user guide_ 
  image
- Move related documentation content to this new topic and add link(s)
- Update `kf-notebook` `Dockerfile` to only enable Kubeflow Pipelines runtime

#### Implementation
 - These changes were tested via a new test which sets the runtime configuration via the environment variable `ELYRA_PROCESSOR_RUNTIMES` and checks various combinations, as well as manual testing to produce the screenshots, etc.
#### Documentation
 - Reviewed output of `make docs`
#### `kf-notebook` container image
 - Build  container image (`make kf-notebook-image`)
 - Run container image (`docker run -it -p 8888:8888 elyra/kf-notebook:dev`)
 - Open JupyterLab UI in web browser
 - In terminal inspect customized content of `$(jupyter --config-dir)/jupyter_elyra_config.py` (should be `c.PipelineProcessorRegistry.runtimes = ['kfp']`)
 - Inspect tiles in JupyterLab launcher window (only KFP pipeline editor tile is rendered)

Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by kevin-bates** [comment]
I'm moving this to draft since there are some documentation updates and potential other changes that are needed.

File: docs/source/user_guide/pipeline-editor-configuration.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+If you are planning to use only one runtime to execute pipelines, you can disable the other runtimes and hide their launcher tiles.
+
+#### Disabling runtimes
+
+Runtimes can be disabled by overriding the Elyra default configuration. 
+
+##### Override default using command line parameters
+
+To disable one or more runtimes, launch JupyterLab with the Elyra-specific `--PipelineProcessorRegistry.runtimes` parameter:
```
Comment (kevin-bates) [inline docs/source/user_guide/pipeline-editor-configuration.md:35]: I find this wording a bit confusing.  It speaks about _disabling_ runtimes, but then references the CLI option to _enable_ a runtime.

I wonder if rephrasing this to something like _Enabling specific runtimes_ and stating within the section contents 
 that this essentially _disables_ the other runtimes might be less confusing.

Comment (ptitzler) [reply docs/source/user_guide/pipeline-editor-configuration.md]: Done. It should be now easier to read.

Comment (kevin-bates) [reply docs/source/user_guide/pipeline-editor-configuration.md]: Looks good - thanks!

**Comment by ptitzler** [comment]
TODO: The run-as dialogs in the python|notebook|r editors need to filter too. Currently all runtime [platforms] are displayed as a selectable option. For example, with only KFP enabled:

image

**Comment by ptitzler** [comment]
TODO: When a runtime is disabled (e.g. 'airflow') and an attempt is made to open a pipeline file for that runtime using the pipeline editor an error is raised

image

Once the user dismisses the dialog the editor remains open (and unusable). We should close it, similar to how it's already done for other scenarios where the pipeline file is not valid.

**Comment by ptitzler** [comment]
TODO: generic VPE: pipeline run and export dialogs allow for selection of a disabled runtime, similar to the issue reported in https://github.com/elyra-ai/elyra/pull/3114#issuecomment-1444104538. In this example an attempt was made to export a generic pipeline with local/kfp enabled and airflow disabled.

image

**Comment by marthacryan** [comment]
> TODO: The run-as dialogs in the python|notebook|r editors need to filter too. Currently all runtime [platforms] are displayed as a selectable option. For example, with only KFP enabled:

> TODO: generic VPE: pipeline run and export dialogs allow for selection of a disabled runtime, similar to the issue reported in [#3114 (comment)](https://github.com/elyra-ai/elyra/pull/3114#issuecomment-1444104538). In this example an attempt was made to export a generic pipeline with local/kfp enabled and airflow disabled.

@ptitzler I think both of these should be fixed?

**Comment by ptitzler** [comment]
> @ptitzler I think both of these should be fixed
Thanks! These are looking good now!

**Comment by ptitzler** [comment]
TODO: 
 - It shoud only be possible to create a runtime environment configuration for enabled runtimes
 - Only runtime configurations for enabled runtimes should be listed in the runtime configurations list

image

**Comment by ptitzler** [comment]
TODO:
- It should only be possible to create catalog connectors for enabled runtimes
   - if no kfp or airflow runtime is enabled, no connectors are supported
   - we need to filter by `metadata.runtime_type` to filter out connectors for disabled runtimes
- Only catalog connectors  for enabled runtimes should be listed in the component catalog list (we need to filter by `metadata.runtime_type`)

image

**Comment by marthacryan** [comment]
> Once the user dismisses the dialog the editor remains open (and unusable). We should close it, similar to how it's already done for other scenarios where the pipeline file is not valid.

Fixed

**Comment by marthacryan** [comment]
> * It should only be possible to create a runtime environment configuration for enabled runtimes
> * Only runtime configurations for enabled runtimes should be listed in the runtime configurations list

Fixed both

**Comment by marthacryan** [comment]
> * It should only be possible to create catalog connectors for enabled runtimes
>   
>   * if no kfp or airflow runtime is enabled, no connectors are supported
>   * we need to filter by `metadata.runtime_type` to filter out connectors for disabled runtimes
> * Only catalog connectors  for enabled runtimes should be listed in the component catalog list (we need to filter by `metadata.runtime_type`)

Done - probably want to add some behavior to say that no runtimes are enabled though

**Comment by kevin-bates** [comment]
> Done - probably want to add some behavior to say that no runtimes are enabled though

Are you able to configure things such that no runtimes are enabled?  And if so, is that a bad thing (assuming things are graceful)?  

I could see it possible to have zero runtimes enabled if a bad value was given in the config, but we should probably weigh the cost of validating that each runtime provided is legit since we may not know what runtimes exist.  I suppose we could check AFTER loading runtimes, and ask "did we load everything that was asked", and, if not, warn about those "configured runtimes" that were not actually loaded.  But, even then, I think we should tolerate no available runtimes if that's doable.

**Comment by ptitzler** [comment]
> > Done - probably want to add some behavior to say that no runtimes are enabled though
> 
> Are you able to configure things such that no runtimes are enabled? And if so, is that a bad thing (assuming things are graceful)?
> 
> I could see it possible to have zero runtimes enabled if a bad value was given in the config, but we should probably weigh the cost of validating that each runtime provided is legit since we may not know what runtimes exist. I suppose we could check AFTER loading runtimes, and ask "did we load everything that was asked", and, if not, warn about those "configured runtimes" that were not actually loaded. But, even then, I think we should tolerate no available runtimes if that's doable.

I did test this scenario by setting `PipelineProcessorRegistry.runtimes` to an invalid string. In the UI all runtimes were disabled and nothing appeared to be broken in an obvious way. The log indicates that the three 'built-in' runtimes are disabled:

```
[I 2023-02-28 14:23:03.462 ElyraApp] Although runtime 'airflow' is installed, it is not in the set of runtimes configured via '--PipelineProcessorRegistry.runtimes' and will not be available.
[I 2023-02-28 14:23:03.935 ElyraApp] Although runtime 'kfp' is installed, it is not in the set of runtimes configured via '--PipelineProcessorRegistry.runtimes' and will not be available.
[I 2023-02-28 14:23:03.996 ElyraApp] Although runtime 'local' is installed, it is not in the set of runtimes configured via '--PipelineProcessorRegistry.runtimes' and will not be available.
```

What it doesn't document though in any message is what `PipelineProcessorRegistry.runtimes` was set to, even though that would be helpful.

**Comment by ptitzler** [comment]
A nit: it appears that the commands are still listed for all runtimes and component catalogs even when the 'associated' runtime is disabled:

image

(Clicking any of the commands does nothing obvious, so there doesn't seem to be any harm.)

**Comment by kevin-bates** [comment]
> A nit: it appears that the commands are still listed for all runtimes and component catalogs even when the 'associated' runtime is disabled:

This may be addressed in my latest commit.

> What it doesn't document though in any message is what PipelineProcessorRegistry.runtimes was set to, even though that would be helpful.

I wonder if instead of displaying ...
```
[I 2023-02-28 14:23:03.462 ElyraApp] Although runtime 'airflow' is installed, it is not in the set of runtimes configured via '--PipelineProcessorRegistry.runtimes' and will not be available.
```
we include the configured runtimes in the message?  Something like this...
```
[I 2023-02-28 14:23:03.462 ElyraApp] Although runtime 'airflow' is installed, it is not in the set of configured runtimes {'kfp', 'local'} and will not be available.
```

**Comment by ptitzler** [comment]
> > What it doesn't document though in any message is what PipelineProcessorRegistry.runtimes was set to, even though that would be helpful.
> 
> I wonder if instead of displaying ...
> 
> ```
> [I 2023-02-28 14:23:03.462 ElyraApp] Although runtime 'airflow' is installed, it is not in the set of runtimes configured via '--PipelineProcessorRegistry.runtimes' and will not be available.
> ```
> 
> we include the configured runtimes in the message? Something like this...
> 
> ```
> [I 2023-02-28 14:23:03.462 ElyraApp] Although runtime 'airflow' is installed, it is not in the set of configured runtimes {'kfp', 'local'} and will not be available.
> ```

I like it because it combines multiple pieces of troubleshooting information! I believe in other places we use [] instead of {} to denote 'sets'. Should we do the same here for consistency?

**Comment by ptitzler** [comment]
> > A nit: it appears that the commands are still listed for all runtimes and component catalogs even when the 'associated' runtime is disabled:

> This may be addressed in my latest commit.

Unfortunately not

**Review by lresende — APPROVED** [review]
LGTM, overall is working fine. One quick question, I built Elyra using `make install-all` and I can't see runtime-specific components like those described in https://elyra.readthedocs.io/en/stable/user_guide/pipeline-components.html#example-custom-components but this is the same behavior from main, so it should be unrelated.

**Comment by ptitzler** [comment]
>  One quick question, I built Elyra using `make install-all` and I can't see runtime-specific components like those described in https://elyra.readthedocs.io/en/stable/user_guide/pipeline-components.html#example-custom-components but this is the same behavior from main, so it should be unrelated.

Maybe re-raise this to question in one of our community channels?

**Review by ptitzler — APPROVED** [review]
LGTM! Thank you! Great feature.

**Comment by ptitzler** [comment]
@marthacryan one odd UI issue. If only `kfp` or `airflow` is enabled the header in the new/edit runtime configuration screen is missing the word `configuration`:

image

image

Everything is fine if both of them are explicitly or implicitly enabled:

image

**Pull request merged**